CODEOWNERS

```diff
@@ -393,6 +393,7 @@
 /scripts/west-commands.yml                @mbolivar
 /scripts/zephyr_module.py                 @tejlmand
 /scripts/valgrind.supp                    @aescolar
+/share/zephyr-package/                    @tejlmand
```

Comment:
not a big fan of adding this top level folder just for the cmake files. Did you think about adding this somewhere else like cmake/, scripts/? What else would end up here?

Reply:
> Did you think about adding this somewhere else like cmake/, scripts/? 

Yes, I did.
And I completely agree, it is not my first choice for location of those files.

The reason I added it here, is the way CMake package search works:
https://cmake.org/cmake/help/latest/command/find_package.html?highlight=find_package#search-procedure

Now, if we locate the package files in `${ZEPHYR_BASE}/cmake/zephyr-package` we would need to write `find_package(Zephyr HINTS $ENV{ZEPHYR_BASE}/cmake/zephyr-package)`, but placing it under `share/zephyr-package/` allows us to simply write `find_package(Zephyr HINTS $ENV{ZEPHYR_BASE})`

Secondly, when ZEPHYR_BASE is not in environment, the search will look for Zephyr in following order:
1) Use Zephyr from current zephyr tree if sample is in-zephyr-tree
2) Use Zephyr from current work-tree, if sample is in-work-tree
3) Use matching Zephyr from user package registry if building out-of-tree.

Some explanation 
1. This would mean that per default,
ZEPHYR_BASE=`/projects/zephyr-project/zephyr/` is used when building
`/projects/zephyr-project/zephyr/samples/hello_world`
2. This could be
ZEPHYR_BASE=`/projects/ncs-project/zephyr/` when building
`/projects/ncs-project/nrf/samples/nrf-sample`
where `/projects/ncs-project/` is the west handled work-tree containing both Zephyr and nRF.


To allow a bit of freedom on how users structure their projects, as well as allowing them a little freedom in naming the Zephyr folder, they should be allowed to also have a  setup like:
```
/projects/my-project/zephyr-my-company/.../
/projects/my-project/my-apps/.../
```

To be able to use `find_package` to locate `/projects/my-project/zephyr-my-company/` in this situation (if user did not run `west zephyr-export` in this work-tree), we would need some matching during the search.

By placing this code under `/share/zephyr-package/cmake` it is ensured that both the `HINTS $ENV{ZEPHYR_BASE}` and the example I just gave can be matched.

```
<prefix>/share/<name>*/cmake/
```
will match when `$ENV{ZEPHYR_BASE}` is used as prefix is `$ENV{ZEPHYR_BASE}`, and name of package is Zephyr, so `zephyr-package` will match the `<name>*` portion.

when having `/projects/my-project/zephyr-my-company/`, the following search pattern will match:
```
<prefix>/<name>*/share/<name>*/cmake/ 
```
as `zephyr-my-company` will match first `<name>*` and `zephyr-package` will match second `<name>*`.

Finally, this approach will also allow users to have multiple Zephyr's installed, such as:
```
/projects/my-project/zephyr  (upstream version ?)
/projects/my-project/zephyr-my-company/.../
/projects/my-project/my-apps/.../
```
however, I still have some more testing to do, to ensure all corner cases are ironed out.
As well as a way to determine preference between `/projects/my-project/zephyr` and `/projects/my-project/zephyr-my-company/` when building `/projects/my-project/my-apps/app-with-zephyr-my-comp_preference`

One more note on this:
Users are allowed to have multiple Zephyrs in their system, and only export one, as example:
```
/projects/zephyr-project/zephyr    # My main Zephyr: west zephyr-export is executed 

/projects/exploration-project/zephyr 
/projects/exploration-project/company-apps

/projects/out-of-tree/company-apps
```

As seen, only `/projects/zephyr-project/zephyr` was exported as CMake package, so whenever building a sample in-side `/projects/zephyr-project/zephyr` or `/projects/zephyr-project/`, then `ZEPHYR_BASE=/projects/zephyr-project/zephyr` will be used.

When building a sample in: `/projects/exploration-project/company-apps` then `ZEPHYR_BASE=/projects/exploration-project/zephyr` will be used, as it is in same work-tree.

When building `/projects/out-of-tree/company-apps` then `ZEPHYR_BASE=/projects/zephyr-project/zephyr` will be used, as this is the only Zephyr package registered in the CMake user package registry.

So with the CMake seach patterns, we basically can choose between:
`lib/arch`, `lib*`, or `share`
```
<prefix>/<name>*/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/
```
and thus  I preferred `share`

(ZEPHYR_BASE=<something> can be used to overrule everything)

Reply:
> Finally, this approach will also allow users to have multiple Zephyr's installed, such as:
> 
> ```
> /projects/my-project/zephyr  (upstream version ?)
> /projects/my-project/zephyr-my-company/.../
> /projects/my-project/my-apps/.../
> ```
>
> however, I still have some more testing to do, to ensure all corner cases are ironed out.
As well as a way to determine preference between /projects/my-project/zephyr and /projects/my-project/zephyr-my-company/ when building /projects/my-project/my-apps/app-with-zephyr-my-comp_preference

and this is also in place now.

Reply:
@nashif as @tejlmand says, this is just not possible if we want to be able to use CMake's package discovery mechanism.